packages/flutter_tools/lib/src/plugins.dart

```diff
@@ -32,7 +32,102 @@ class Plugin {
     this.pluginClass,
   });
 
+  /// Parses [Plugin] specification from the provided pluginYaml.
+  ///
+  /// This currently supports two formats. V1 (Legacy) and V2 (Multi-platform).
+  /// Example V1 format. Which is deprecated.
+  /// flutter:
+  ///  plugin:
+  ///    androidPackage: io.flutter.plugins.sample
+  ///    iosPrefix: FLT
+  ///    macosPrefix: FLT
+  ///    pluginClass: SamplePlugin
+  ///
+  /// Example V2 format.
+  /// flutter:
+  ///  plugin:
+  ///    platforms:
+  ///      android:
+  ///        package: io.flutter.plugins.sample
+  ///        pluginClass: SamplePlugin
+  ///      ios:
+  ///        classPrefix: FLT
+  ///        pluginClass: SamplePlugin
+  ///      macos:
+  ///        classPrefix: FLT
+  ///        pluginClass: SamplePlugin
   factory Plugin.fromYaml(String name, String path, dynamic pluginYaml) {
+    // TODO(kaushikiska): Fail if both old and new formats are present.
+    if (pluginYaml != null && pluginYaml['platforms'] != null) {
+      return Plugin._fromMultiPlatformYaml(name, path, pluginYaml);
+    } else {
+      return Plugin._fromLegacyYaml(name, path, pluginYaml);
+    }
+  }
+
+  factory Plugin._fromMultiPlatformYaml(String name, String path, dynamic pluginYaml) {
+    assert (pluginYaml != null && pluginYaml['platforms'] != null,
+            'Invalid multi-platform plugin specification.');
+    final dynamic platforms = pluginYaml['platforms'];
+
+    final Set<String> pluginClassNames = {};
+    String androidPackage;
+    String iosPrefix;
+    String macosPrefix;
+    String pluginClass;
+
+    if (platforms['android'] != null) {
+      final dynamic androidConfig = platforms['android'];
+      androidPackage = androidConfig['package'];
+      if (androidConfig['pluginClass'] != null) {
+        pluginClassNames.add(androidConfig['pluginClass']);
+      } else {
+        throw Exception('"pluginClass" needs to be specified for android platform');
```

Comment:
might be worth creating a new exception type for failures relating to parsing a plugin. I haven't seen any custom exceptions in the `flutter_tools` project so I stuck with `Exception`.